From 44ca211951a4214950742d87e535417e02c27e3e Mon Sep 17 00:00:00 2001
From: Andrew Clemons <andrew.clemons@gmail.com>
Date: Mon, 25 Sep 2017 14:00:37 +1300
Subject: [PATCH] Mark projects which have unsupported dependencies as
 unsupported

---
 libexec/functions.d/cmdfunctions.sh | 26 ++++++++++++++++++++------
 1 file changed, 20 insertions(+), 6 deletions(-)

diff --git a/libexec/functions.d/cmdfunctions.sh b/libexec/functions.d/cmdfunctions.sh
index 2df2708..f30c821 100755
--- a/libexec/functions.d/cmdfunctions.sh
+++ b/libexec/functions.d/cmdfunctions.sh
@@ -123,22 +123,36 @@ function build_command
         STATUS[$todo]='removed'
       else
         missingdeps=()
+        unsupporteddeps=()
         for dep in ${DIRECTDEPS[$todo]}; do
           if [ "${STATUS[$dep]}" != 'ok' ] && [ "${STATUS[$dep]}" != 'updated' ]; then
-            missingdeps+=( "$dep" )
+            if [ "${STATUS[$dep]}" = 'unsupported' ] ; then
+              unsupporteddeps+=( "$dep" )
+            else
+              missingdeps+=( "$dep" )
+            fi
           fi
         done
-        if [ "${#missingdeps[@]}" = '0' ]; then
+        if [ "${#missingdeps[@]}" = '0' ] && [ "${#unsupporteddeps[@]}" = '0' ] ; then
           build_item_packages "$todo"
         else
           log_error "Cannot build ${todo}."
-          if [ "${#missingdeps[@]}" = '1' ]; then
+
+          if [ "${#missingdeps[@]}" = '1' ] && [ "${#unsupportedeps[@]}" = '0' ] ; then
             STATUSINFO[$todo]="Missing dependency: ${missingdeps[0]}"
+          elif [ "${#missingdeps[@]}" = '0' ] && [ "${#unsupportedeps[@]}" = '1' ] ; then
+            STATUSINFO[$todo]="Unsupported dependency: ${unsupporteddeps[0]}"
+          else
+            STATUSINFO[$todo]="Missing dependencies:\n$(printf '  %s\n' "${missingdeps[@]}")\nUnsupported dependencies:\n$(printf '  %s\n' "${unsupporteddeps[@]}")"
+          fi
+
+          if [ "${#unsupporteddeps[@]}" = '0' ] ; then
+            STATUS[$todo]='aborted'
+            log_itemfinish "$todo" "aborted" '' "${STATUSINFO[$todo]}"
           else
-            STATUSINFO[$todo]="Missing dependencies:\n$(printf '  %s\n' "${missingdeps[@]}")"
+            STATUS[$todo]='unsupported'
+            log_itemfinish "$todo" "unsupported" '' "${STATUSINFO[$todo]}"
           fi
-          STATUS[$todo]='aborted'
-          log_itemfinish "$todo" "aborted" '' "${STATUSINFO[$todo]}"
         fi
       fi
     done
